src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs

```diff
@@ -271,7 +271,7 @@ internal override bool TryGetSpeculativeSemanticModelCore(SyntaxTreeSemanticMode
         {
             if (MemberSymbol is SynthesizedRecordConstructor primaryCtor &&
                 primaryCtor.GetSyntax() is RecordDeclarationSyntax recordDecl &&
-                Root.FindToken(position).Parent?.AncestorsAndSelf().OfType<PrimaryConstructorBaseTypeSyntax>().FirstOrDefault() == recordDecl.PrimaryConstructorBaseType)
+                Root.FindToken(position).Parent?.AncestorsAndSelf().OfType<PrimaryConstructorBaseTypeSyntax>().FirstOrDefault() == recordDecl.PrimaryConstructorBaseTypeIfClass)
```

Comment:
>Root.FindToken(position).Parent?.AncestorsAndSelf().OfType<PrimaryConstructorBaseTypeSyntax>().FirstOrDefault() == recordDecl.PrimaryConstructorBaseTypeIfClass [](http://example.com/codeflow?start=16&length=159)

I am assuming this method is never called for a record struct, otherwise it should never return true, even in error scenarios. It might be worth adding an assert at the top of the method that ```MemberSymbol``` is not a primary constructor of a record struct. Also, I think we need to add a test that tries to get the speculative semantic model for a record struct at position of the first base when it has and doesn't have arguments. BTW, it looks like the is a bug, the following code compiles without any error:
 ```
public interface I1
{
    void M01();
}

record struct R(int X) : I1(X)
{
    void I1.M01() {}
}
```

Reply:
There is also a bug for a regular record:
```
record R(int X) : I1()
{
    void I1.M01() {}
}
```
No error.

Reply:
Added test `BaseArguments_Speculation`. That test will reach the body of this conditional when using `record` instead of `record struct`. With `record struct` we don't even get to the `TryGetSpeculativeSemanticModelCore` method.
I think it's because of the logic in `SyntaxTreeSemanticModel.GetMemberModel` which handles `SyntaxKind.RecordDeclaration` but not `SyntaxKind.RecordStructDeclaration`

---
In reply to: [617049655](https://github.com/dotnet/roslyn/pull/52702#discussion_r617049655) [](ancestors = 617049655)

Reply:
That matches my expectations, that is why I am suggesting to add an assert to ensure the method (not the body of the if) is not reachable for a record struct